examples/retrieval/retrieval.cpp

```diff
@@ -0,0 +1,358 @@
+#include "common.h"
+#include "llama.h"
+
+#include <algorithm>
+#include <fstream>
+
+struct retrieval_params {
+    std::vector<std::string> context_files;// context files to embed
+    int32_t chunk_size            = 64;    // chunk size for context embedding
+    std::string chunk_separator   = "\n";  // chunk separator for context embedding
+};
+
+static void retrieval_params_print_usage(int argc, char** argv, gpt_params & gpt_params, retrieval_params & params) {
+    fprintf(stderr, "usage: retrieval [options]\n");
+    fprintf(stderr, "options:\n");
+    printf("  --context-files FNAME1 FNAME2...\n");
+    printf("                        files containing context to embed.\n");
+    printf("  --chunk-size N        minimum length of embedded text chunk (default:%d)\n", params.chunk_size);
+    printf("  --chunk-separator STRING\n");
+    printf("                        string to separate chunks (default: \"\\n\")\n");
+    gpt_print_usage(argc, argv, gpt_params);
+}
+
+static void retrieval_params_parse(int argc, char ** argv, gpt_params & gpt_params, retrieval_params & retrieval_params) {
+    int i = 1;
+    std::string arg;
+    while(i < argc) {
+        arg = argv[i];
+        bool invalid_gpt_param = false;
+        if(gpt_params_find_arg(argc, argv, gpt_params, i, invalid_gpt_param)) {
+            if (invalid_gpt_param) {
+                fprintf(stderr, "error: invalid argument: %s\n", arg.c_str());
+                retrieval_params_print_usage(argc, argv, gpt_params, retrieval_params);
+                exit(1);
+            }
+            // option was parsed by gpt_params_find_arg
+        } else if (arg == "--context-files") {
+            if (++i >= argc) {
+                fprintf(stderr, "error: missing argument for --context-files\n");
+                retrieval_params_print_usage(argc, argv, gpt_params, retrieval_params);
+                exit(1);
+            }
+            while(true) {
+                std::ifstream file(argv[i]);
+                if (!file) {
+                    fprintf(stderr, "error: failed to open file '%s'\n", argv[i]);
+                    retrieval_params_print_usage(argc, argv, gpt_params, retrieval_params);
+                    exit(1);
+                }
+                // store the external file name in params
+                retrieval_params.context_files.push_back(argv[i]);
+                if (i + 1 >= argc || argv[i + 1][0] == '-') {
```

Comment:
This logic won't work if a file name starts with dash (`-`) character

Here we should specify multiple files by calling `--context-files` multiple times, for example:

```
--context-files file1.txt --context-files file2.txt ...
```

This behavior should also be written in README, so users can quickly understand how to use it.